### PR TITLE
fix 'route delete' on 2.5

### PR DESCRIPTION
**Pull request by vktg**
- [ ] Redmine Issue: https://redmine.pfsense.org/issues/10001
- [ ] Ready for review

All code that uses “delete route” must use the full format with GW, i.e.

incorrect:
```
# route delete -inet 5.5.5.0/24
route: route has not been found
delete net 5.5.5.0 fib 0: not in table
```

correct:
```
# route delete -inet 5.5.5.0/24 10.1.0.254
delete net 5.5.5.0: gateway 10.1.0.254
```

no such issue with 2.4.4-p3 and 2.4.5
**it seems FreeBSD 12 issue**

Same issue: https://redmine.pfsense.org/issues/9969

### CODE REVIEW

File: src/etc/inc/gwlb.inc [hunk 1]
```diff
@@ -2026,17 +2026,19 @@ function save_gateway($gateway_settings, $realid = "") {
 		isset($a_gateway_item[$realid]["nonlocalgateway"])) {
 		$realif = get_real_interface($a_gateway_item[$realid]['interface']);
 		$inet = (!is_ipaddrv4($a_gateway_item[$realid]['gateway']) ? "-inet6" : "-inet");
-		$cmd = "/sbin/route delete $inet " . escapeshellarg($a_gateway_item[$realid]['gateway']) . " -iface " . escapeshellarg($realif);
+		$rgateway = $a_gateway_item[$realid]['gateway'];
+		$cmd = "/sbin/route delete $inet " . escapeshellarg($a_gateway_item[$realid]['gateway']) . " -iface " . escapeshellarg($realif . " " . escapeshellarg($rgateway));
```
Comment (jim-p) [inline src/etc/inc/gwlb.inc]: `escapeshellarg($realif . " " . escapeshellarg($rgateway))` -- Is that intentional? `$rgateway` will be escaped and then the whole concatenated string escaped again. Looks to me like the closing paren for the first `escapeshellarg` needs moved to just after `$realif`, rather than doing it around both.

File: src/etc/inc/gwlb.inc [hunk 2]
```diff
@@ -2125,7 +2127,7 @@ function gateway_set_enabled($name, $enabled) {
 					$cgw = getcurrentdefaultgatewayip('inet6');
 				}
 				if ($gateway['gateway'] == $cgw) {
-					mwexec("/sbin/route delete -{$inet} default");
+					mwexec("/sbin/route delete -{$inet} default {$cgw}");
```
Comment (jim-p) [inline src/etc/inc/gwlb.inc]: Needs `escapeshellarg()`.

File: src/etc/inc/system.inc [hunk 3]
```diff
@@ -202,10 +202,11 @@ function system_resolvconf_generate($dynupdate = false) {
 				route_add_or_change("-host {$inet6}{$dnsserver} {$gatewayip}");
 			} else {
 				/* Remove old route when disable gw */
-				mwexec("/sbin/route -q delete -host {$inet6}{$dnsserver}");
+				$gatewayip = exec("/sbin/route -n show {$dnsserver} | grep gateway | cut -d' ' -f6");
```
Comment (jim-p) [inline src/etc/inc/system.inc]: * Use `route -n get`, rather than `show` to be consistent with other uses elsewhere in our code.
* Use the full path to binaries.
* That `cut` is inaccurate. Swap both `grep` and `cut` out for a single `awk`:
```/sbin/route -n get {$dnsserver} | /usr/bin/awk '/gateway:/ {print $2;}'```

File: src/etc/inc/system.inc [hunk 4]
```diff
@@ -202,10 +202,11 @@ function system_resolvconf_generate($dynupdate = false) {
...
+				mwexec("/sbin/route -q delete -host {$inet6}{$dnsserver} {$gatewayip}");
```
Comment (jim-p) [inline src/etc/inc/system.inc]: Use `escapeshellarg()` on `$gatewayip`.

File: src/usr/local/www/system_gateways.php [hunk 5]
```diff
@@ -142,8 +142,9 @@ function delete_gateway_item($id) {
 	    !isset($a_gateways[$id]['disabled']) &&
 	    isset($a_gateways[$id]['isdefaultgw'])) {
 		$inet = (!is_ipaddrv4($a_gateways[$id]['gateway']) ? '-inet6' : '-inet');
+		$rgateway = $a_gateways[$id]['gateway'];
```
Comment (jim-p) [inline src/usr/local/www/system_gateways.php]: Why set `$rgateway` when it's only used once in these instances? Why not just use `$a_gateways[$id]['gateway']` directly?

File: src/usr/local/www/system_gateways.php [hunk 6]
```diff
@@ -142,8 +142,9 @@ function delete_gateway_item($id) {
...
 		file_put_contents("/dev/console", "\n[".getmypid()."] DEL_GW, route= delete {$inet} default");
-		mwexec("/sbin/route delete {$inet} default");
+		mwexec("/sbin/route delete {$inet} default {$rgateway}");
```
Comment (jim-p) [inline src/usr/local/www/system_gateways.php]: Needs `escapeshellarg()`.

File: src/usr/local/www/system_routes_edit.php [hunk 7]
```diff
@@ -199,8 +199,8 @@
 				foreach ($delete_targets as $dts) {
 					if (is_ipaddrv6($dts)) {
 						$family = "-inet6";
-					}
```
Comment (jim-p) [inline src/usr/local/www/system_routes_edit.php]: Why was this removed? Looks like it would introduce a syntax error here.

File: src/usr/local/www/system_routes_edit.php [hunk 8]
```diff
@@ -199,8 +199,8 @@
...
-					$toapplylist[] = "/sbin/route delete {$family} {$dts}";
+					$rgateway = exec("/sbin/route -n show {$dts} | grep gateway | cut -d' ' -f6");
```
Comment (jim-p) [inline src/usr/local/www/system_routes_edit.php]: See previous note about `grep`/`cut` vs `awk`.

File: src/usr/local/www/system_routes_edit.php [hunk 9]
```diff
@@ -199,8 +199,8 @@
...
+					$toapplylist[] = "/sbin/route delete {$family} {$dts} {$rgateway}";
```
Comment (jim-p) [inline src/usr/local/www/system_routes_edit.php]: Needs `escapeshellarg()`.

File: src/usr/local/www/vpn_ipsec.php [hunk 10]
```diff
@@ -197,7 +197,8 @@
 	} else if (isset($delbtn)) {
 		/* remove static route if interface is not WAN */
 		if ($a_phase1[$delbtn]['interface'] <> "wan") {
-			mwexec("/sbin/route delete -host {$a_phase1[$delbtn]['remote-gateway']}");
+			$rgateway = exec("/sbin/route -n show {$a_phase1[$delbtn]['remote-gateway']} | grep gateway | cut -d' ' -f6");
```
Comment (jim-p) [inline src/usr/local/www/vpn_ipsec.php]: See previous note about `grep`/`cut` vs `awk`.

File: src/usr/local/www/vpn_ipsec.php [hunk 11]
```diff
@@ -197,7 +197,8 @@
...
+			mwexec("/sbin/route delete -host {$a_phase1[$delbtn]['remote-gateway']} {$rgateway}");
```
Comment (jim-p) [inline src/usr/local/www/vpn_ipsec.php]: Needs `escapeshellarg()`.

File: src/usr/local/www/vpn_ipsec_phase1.php [hunk 12]
```diff
@@ -468,7 +468,8 @@
 		/* the ipsec_configure() handles adding the route */
 		if ($pconfig['interface'] <> "wan") {
 			if ($old_ph1ent['remote-gateway'] <> $pconfig['remotegw']) {
-				mwexec("/sbin/route delete -host {$old_ph1ent['remote-gateway']}");
+                                $rgateway = exec("/sbin/route -n show {$old_ph1ent} | grep gateway | cut -d' ' -f6");
```
Comment (jim-p) [inline src/usr/local/www/vpn_ipsec_phase1.php]: See previous note about `grep`/`cut` vs `awk`.

File: src/usr/local/www/vpn_ipsec_phase1.php [hunk 13]
```diff
@@ -468,7 +468,8 @@
...
+				mwexec("/sbin/route delete -host {$old_ph1ent['remote-gateway']} {$rgateway}");
```
Comment (jim-p) [inline src/usr/local/www/vpn_ipsec_phase1.php]: Needs `escapeshellarg()`.

File: src/etc/inc/gwlb.inc (hunk 1 above)
Comment (vktg) [reply src/etc/inc/gwlb.inc]: typo, fixed

File: src/etc/inc/gwlb.inc (hunk 2 above)
Comment (vktg) [reply src/etc/inc/gwlb.inc]: done

File: src/etc/inc/system.inc (hunk 3 above)
Comment (vktg) [reply src/etc/inc/system.inc]: done

File: src/etc/inc/system.inc (hunk 4 above)
Comment (vktg) [reply src/etc/inc/system.inc]: done

File: src/usr/local/www/system_gateways.php (hunk 5 above)
Comment (vktg) [reply src/usr/local/www/system_gateways.php]: done

File: src/usr/local/www/system_gateways.php (hunk 6 above)
Comment (vktg) [reply src/usr/local/www/system_gateways.php]: done

File: src/usr/local/www/system_routes_edit.php (hunk 7 above)
Comment (vktg) [reply src/usr/local/www/system_routes_edit.php]: typo, fixed

File: src/usr/local/www/system_routes_edit.php (hunk 8 above)
Comment (vktg) [reply src/usr/local/www/system_routes_edit.php]: done

File: src/usr/local/www/system_routes_edit.php (hunk 9 above)
Comment (vktg) [reply src/usr/local/www/system_routes_edit.php]: done

File: src/usr/local/www/vpn_ipsec.php (hunk 10 above)
Comment (vktg) [reply src/usr/local/www/vpn_ipsec.php]: done

File: src/usr/local/www/vpn_ipsec.php (hunk 11 above)
Comment (vktg) [reply src/usr/local/www/vpn_ipsec.php]: done

File: src/usr/local/www/vpn_ipsec_phase1.php (hunk 12 above)
Comment (vktg) [reply src/usr/local/www/vpn_ipsec_phase1.php]: done

File: src/usr/local/www/vpn_ipsec_phase1.php (hunk 13 above)
Comment (vktg) [reply src/usr/local/www/vpn_ipsec_phase1.php]: done
